packages/runtime/container-runtime/src/gc/gcTelemetryTracker.ts

```diff
@@ -0,0 +1,353 @@
+	completedGCRuns: number;
+	isTombstoned: boolean;
+	lastSummaryTime?: number;
+	fromId?: string;
+	viaHandle?: boolean;
+}
+
+/**
+ * Encapsulates the logic that tracks the various telemetry logged by the Garbage Collector. There are 4 types of
+ * telemetry logged:
+ * 1. inactiveObject telemetry - When an inactive node is used - A node that has been unreferenced for inactiveTimeoutMs.
+ * 2. sweepReadyObject telemetry - When a sweep ready node is used - A node that has been unreferenced for sweepTimeoutMs.
+ * 3. Tombstone telemetry - When a tombstoned node is used - A node that that has been marked as tombstone.
+ * 4. Sweep / deleted telemetry - When a node is detected as sweep ready in the sweep phase.
+ * 5. Unknown outbound reference telemetry - When a node is referenced but GC is not explicitly notified of it.
+ */
+export class GCTelemetryTracker {
+	// Keeps track of unreferenced events that are logged for a node. This is used to limit the log generation to one
+	// per event per node.
+	private readonly loggedUnreferencedEvents: Set<string> = new Set();
+	// Queue for unreferenced events that should be logged the next time GC runs.
+	private pendingEventsQueue: IUnreferencedEventProps[] = [];
+
+	constructor(
+		private readonly mc: MonitoringContext,
+		private readonly configs: Pick<
+			IGarbageCollectorConfigs,
+			"inactiveTimeoutMs" | "sweepTimeoutMs"
+		>,
+		private readonly isSummarizerClient: boolean,
+		private readonly gcTombstoneEnforcementAllowed: boolean,
+		private readonly createContainerMetadata: ICreateContainerMetadata,
+		private readonly getNodeType: (nodeId: string) => GCNodeType,
+		private readonly getNodeStateTracker: (
+			nodeId: string,
+		) => UnreferencedStateTracker | undefined,
+		private readonly getNodePackagePath: (
+			nodePath: string,
+		) => Promise<readonly string[] | undefined>,
+	) {}
+
+	/**
+	 * Returns whether an event should be logged for a not that isn't active anymore. Some scenarios where we won't log:
+	 * 1. When a DDS is changed or loaded. The corresponding data store's event will be logged instead.
+	 * 2. An event is logged only once per container instance per event per node.
+	 */
+	private shouldLogNonActiveEvent(
+		nodeId: string,
+		nodeType: GCNodeType,
+		usageType: NodeUsageType,
+		nodeStateTracker: UnreferencedStateTracker,
+	) {
+		if (nodeStateTracker.state === UnreferencedState.Active) {
+			return false;
+		}
+
+		// For sub data store (DDS) nodes, if they are changed or loaded, its data store will also be changed or loaded,
+		// so skip logging to make the telemetry less noisy.
+		if (nodeType === GCNodeType.SubDataStore && usageType !== "Revived") {
+			return false;
+		}
+		if (nodeType === GCNodeType.Other) {
+			return false;
+		}
+
+		const uniqueEventId = `${nodeStateTracker.state}-${nodeId}-${usageType}`;
+		if (this.loggedUnreferencedEvents.has(uniqueEventId)) {
+			return false;
+		}
+		this.loggedUnreferencedEvents.add(uniqueEventId);
```

Comment:
To me this is a surprising side effect of a function called `should...`

Reply:
Good point. I should move this to the call site.